File: src/Patterns/HalfDiamond.java
/*
Write a program to print N number of rows for Half Diamond pattern using stars and numbers
Note : There are no spaces between the characters in a single line.


Input Format :
A single integer: N
Output Format :
Required Pattern
Constraints :
0 <= N <= 50
Sample Input 1 :
3
Sample Output 1 :
*
*1*
*121*
*12321*
*121*
*1*
*
Sample Input 2 :
 5
Sample Output 2 :
*
*1*
*121*
*12321*
*1234321*
*123454321*
*1234321*
*12321*
*121*
*1*
*
 */
package Patterns;

import java.util.Scanner;

public class HalfDiamond {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n   = sc.nextInt();
        System.out.println("*");
        for(int i = 1; i <= n; i++) {
            System.out.print("*");
            for(int j = 1;j<=i; j++){
                System.out.print(j);

            }
            for(int k=2; k<=i; k++) {
                System.out.print(i-k+1);

            }
            System.out.print("*");
            System.out.println();
        }

        for(int i=1; i <n; i++) {
            System.out.print("*");
            for(int j=1;j<=n-i; j++){
                System.out.print(j);
            }
            for(int k=i; k<=n-2; k++) {
                System.out.print(n-k-1);
            }
            System.out.print("*");
            System.out.println();
        }
        System.out.println("*");
    }
}

File: src/Patterns/RecursivePattern/HalfDiamond.py
def main():
    n = int(input())
    p=n
    print("*")
    pattern(n,p)
    dpattern(n-1,p-1)
    print("*")

def pattern(n,p):
    if(n<=0):
        return
    print("*",end="")
    for i in range (1,p-n+2):
        print(i,end="")
    
    if(n<p):
        for i in range(1,p-n+1):
            print(p-n+1-i,end="")
    
    print("*")


    pattern(n-1,p)

def dpattern(n,p):
    if(n<=0):
        return
    print("*",end="")
    for i in range(1,n+1):
        print(i,end="")
    if(n>1):
        for j in range(n-1,0,-1):
            print(j,end="")
    print("*")
    dpattern(n-1,p)

main()